Improve patch and dev version support

brodybits opened the pull request:
<!--
Please make sure the checklist boxes are all checked before submitting the PR. The checklist
is intended as a quick reference, for complete details please see our Contributor Guidelines:

http://cordova.apache.org/contribute/contribute_guidelines.html

Thanks!
-->

### Platforms affected

All

### What does this PR do?

- fix usage messages
- prepare iOS `CDVAvailability.h` after updating JS in git (more sensical order)
- do not update iOS `CDVAvailability.h` if version number ends with non-numerical character, needed to avoid build error in case of marking -dev version
- workaround for local grunt issue in cordova-js from PR #174 (hope to look into a more "real" solution, someday, will probably comment more in #174)
- Updates to improve support for patch releases (see rationale in PR #176)
  - `prepare-platform-release-branch` add options:
    - `[-b <platform branch name>]`
    - `[-js <cordova-js branch or tag name>]`
  - `copy-js` add option:
    - `[-js <cordova-js branch or tag name>]`
- Add `--tag-only` option to `tag-release ` command (shows the `git push` commands that can be done manually)
- EXTRA GIT WORKAROUND solution to do `git checkout -- package.json`, needed for strange cases with cordova-osx (4.0.2-dev) & cordova-windows
- show slightly modified commit message in case version is updated in package.json
- add `--pre` option to `prepare-platform-release-branch` to allow specific prefix such as JIRA or GH number in the commit messages
- remove some extra info from commit messages
- other minor fixes

This proposal supersedes and resolves #174 (GH-174).

### What testing has been done on this change?

GH-174 was successfully used for a number of patch releases including:
- apache/cordova-browser#55 - CB-14230 patch release 5.0.4
- apache/cordova-osx#51 - CB-14228 release 4.0.2
- apache/cordova-windows#287 - CB-14226 cordova-windows@6.0.1 (July 2018 patch release)
- apache/cordova-ios#381 - CB-14217 cordova-ios 4.5.5 patch release (July 2018)
- apache/cordova-android#467 - CB-14203 patch release
- apache/cordova-osx#50 - CB-14145 resolve npm audit issues in patch fix
- apache/cordova-windows#281 - CB-14145 resolve npm audit issues in patch fix
- apache/cordova-ios#379 - CB-14145 resolve npm audit issues in 4.5.x patch fix
- apache/cordova-android#451 - CB-14145: cordova-common update to resolve npm audit & other updates in patch release
- apache/cordova-browser#53 - CB-14145 resolve npm audit issues in patch release

(updated usage as needed for the patch releases is shown in most of these PRs)

Additional changes to add the `--pre` option & better deal with -dev version are demonstrated working in:

- apache/cordova-ios#387 - CB-14244 start 5.0.0-dev in master branch (includes updated command usage that was used there)

### Checklist

- ~~[Reported an issue](http://cordova.apache.org/contribute/issues.html) in the JIRA database~~
- ~~Commit message follows the format: "CB-3232: (android) Fix bug with resolving file paths", where CB-xxxx is the JIRA ID & "android" is the platform affected.~~
- ~~Added automated test coverage as appropriate for this change.~~

janpio commented on docs/platforms-release-process.md:
```diff
@@ -260,13 +260,19 @@ Create and prepare your release branch by using `coho prepare-platform-release-b
-    coho repo-status -r android -b master -b 5.0.x
+    coho repo-status -r android -b master -b 7.1.x
+
+or use git tool to verify manually.
```
"git tool"? How? Should I check a commit message`or what exactly?

janpio replied:
Then lets formulate this more general:

```suggestion
or verify manually (e.g. by looking at the git commit messages).
```

janpio commented on docs/platforms-release-process.md:
```diff
@@ -357,25 +363,31 @@ Create a JIRA issue for it, and mark it as a blocker.
-    coho tag-platform-release --version 5.0.0 -r android
+    coho tag-platform-release --version 7.1.0 -r android
+
+This command also tags `cordova-js` with `android-7.1.0` and pushes it.
```
What does this mean for the person running it? What are the consequences so that this has to be mentioned here?

janpio commented on docs/platforms-release-process.md:
```diff
@@ -357,25 +363,31 @@ Create a JIRA issue for it, and mark it as a blocker.
+
+This command also tags `cordova-js` with `android-7.1.0` and pushes it.
+
+To tag without automatic push:
```
Why would one want to use that instead of automatic push?

brodybits replied:
I personally do not like automatic push. I would rather review my most recent changes before pushing to GitHub.

janpio commented on src/gitutil.js:
```diff
@@ -137,6 +137,10 @@ exports.pendingChangesExist = function * () {
 exports.gitCheckout = function * (branchName) {
     var curBranch = yield gitutil.retrieveCurrentBranchName(true);
     if (curBranch !== branchName) {
+        // EXTRA WORKAROUND SOLUTION for package.json,
```
Is there an issue somewhere that describes what problem is happening here?

brodybits replied:
Added TODO comments to raise issues

janpio commented on src/platform-release.js:
```diff
@@ -139,10 +139,19 @@ function * updateJsSnapshot (repo, version, commit) {
+                    yield gitutil.gitCheckout(branch);
+                    yield executil.execHelper(executil.ARGS('npm install'), false, true); // WORKAROUND PART 1 for local grunt issue in cordova-js
                     yield executil.execHelper(executil.ARGS('grunt compile:' + repo.id + ' --platformVersion=' + version), false, true);
+                    shelljs.rm('-fr', 'node_modules'); // WORKAROUND PART 2 for local grunt issue in cordova-js
```
Problem description in issue somewhere?

brodybits replied:
added TODO comments to raise issues

janpio commented on src/platform-release.js:
```diff
@@ -197,64 +210,87 @@ exports.prepareReleaseBranchCommand = function * () {
 
     var repos = flagutil.computeReposFromFlag(argv.r);
+
+    // XXX TBD ???:
```
?

janpio commented:
> workaround for local grunt issue in cordova-js from PR #174

What does this mean? Issue in a PR?

> This proposal supersedes and resolves #174 (GH-174).

#174 is already closed though!?

> ... needed for strange cases with cordova-osx (4.0.2-dev) & cordova-windows

What strange cases?

> GH-174 was successfully used for a number of patch releases including:
> ...

What does this mean in the context of this PR? I don't get the connection.

janpio commented:
In general: I think this should not be 1 PR, but several that each do less things and have a clear problem description in the PR description (or link to an issue somewhere). Right now this is too much to understand and assign which problem is solved by which line of change.

brodybits commented:
Thanks @janpio for reviewing. I have been keeping this branch to solve some mostly basic problems with making branch releases & dev versions. Unfortunately I haven't had so much time to go through and document the exact problems in better detail.

I would like to finish with what I think are some urgent tasks remaining then can go through, document in more detail, and raise in separate PRs for review.

Does that sound OK?

janpio commented:
Yep, not in a hurry - just wanted to make sure that all open questions to the changes being made are answered and we won't be mad in a year or so about that massive commit without all the necessary context ;)

janpio commented:
Hey @brodybits, after going through the release process documentation again to try to make sense of it (and possibly unify it for all repos), I think most of the actual changes on coho you suggest here make sense and might be worth splitting into individual PRs to be merged as soon as possible.

(I am not sure about the `package.json` and grunt stuff)

Do you possibly have the time to fix the tiny merge conflict here and then split this into individual PRs?

brodybits commented on src/platform-release.js:
```diff
@@ -228,13 +228,15 @@ exports.prepareReleaseBranchCommand = function * () {
                 yield executil.execHelper(executil.ARGS('git checkout -b ' + branchName));
             }
 
+            print(repo.repoName + ': Update JS snapshot version for VERSION "' + version + '" on branch "' + branchName + '".');
```
Both changes from this commit (aefb169daa57f3931b08aa8ea63866dcd58117ba) were clobbered by a4110226b5e5ad9f36a1782dd8ca972870c0f0cb and will likely go away.

brodybits commented on src/platform-release.js:
```diff
@@ -228,13 +228,15 @@ exports.prepareReleaseBranchCommand = function * () {
 
             if (platform === 'ios') {
                 // Updates version in CDVAvailability.h file
+                print(repo.repoName + ': Update CDVAvailability.h for VERSION to "' + version + '" on branch "' + branchName + '".');
```
Both changes from this commit (aefb169daa57f3931b08aa8ea63866dcd58117ba) were clobbered by a4110226b5e5ad9f36a1782dd8ca972870c0f0cb and will likely go away.

brodybits commented:
I just merged in the changes from `master`, updated some XXX TBD comments, added some TODO comments, and did `eslint --fix`. I reverted some changes that I made for iOS to avoid a merge conflict for now.

I also rebased the changes with 7b54e2752305a3b833885bb94fb915b0c038a3c8 and aefb169daa57f3931b08aa8ea63866dcd58117ba removed: <https://github.com/apache/cordova-coho/compare/master...brodybits:pr-188-rebased>

Unfortunately I cannot promise when I will get a chance to split the changes as I am falling behind on some other commitments.
